torch/_inductor/fx_passes/b2b_gemm.py

```diff
@@ -0,0 +1,206 @@
+    grid=b2b_gemm_grid,
+    debug=False,
+    source=r"""
+{{def_kernel("A", "B", "C")}}
+
+
+    # B2B_GEMM_TRITON_ENTRANCE
+    # dram load/store estimations
+    #   (A @ B) @ C
+    #   M * N, N * O, O * P
+    #   baseline
+    #     load = M * N + N * O + M * O + O * P
+    #     store = M * O + M * P
+    #   gemm
+    #     load = M * N + M / m * (N * O + O * P)
+    #     store = M * P
+    M = {{size("A", 0)}}
+    # N = {{size("A", 1)}}
+    O = {{size("C", 0)}}
+    # P = {{size("C", 1)}}
+
+    stride_am = {{stride("A", 0)}}
+    stride_an = {{stride("A", 1)}}
+    stride_bn = {{stride("B", 0)}}
+    stride_bo = {{stride("B", 1)}}
+    stride_co = {{stride("C", 0)}}
+    stride_cp = {{stride("C", 1)}}
+
+    # A's row block for this thread
+    row_block_id = tl.program_id(axis=0)
+
+    # divide B's columns (and C's rows)
+    num_col_block = tl.cdiv(O, COL_BLOCK_SIZE)
+
+    # offsets (TODO: handle the non-divisible case)
+    offs_row = row_block_id * ROW_BLOCK_SIZE + tl.arange(0, ROW_BLOCK_SIZE)
+    offs_col = tl.arange(0, COL_BLOCK_SIZE)  # to be updated in the loop
+
+    # accumulator
+    acc = tl.zeros((ROW_BLOCK_SIZE, P), dtype=tl.float16)
+
+    a_ptrs = A + (offs_row[:, None] * stride_am + tl.arange(0, N)[None, :] * stride_an)
+    a = tl.load(a_ptrs)
+
+    for _ in range(num_col_block):
+
+        b_ptrs = B + (tl.arange(0, N)[:, None] * stride_bn + offs_col[None, :] * stride_bo)
+        b = tl.load(b_ptrs)
+
+        c_ptrs = C + (offs_col[:, None] * stride_co + tl.arange(0, P)[None, :] * stride_cp)
+        c = tl.load(c_ptrs)
+
+        # computation (TODO: floating point errors)
+        acc += tl.dot(tl.dot(a, b, out_dtype=tl.float16), c, out_dtype=tl.float16)
+
+        # update offsets
+        offs_col += COL_BLOCK_SIZE
+
+    # store
+    idx_m = offs_row[:, None]
+    idx_p = tl.arange(0, P)[None, :]
+    mask = (idx_m < M) & (idx_p < P)
+
+    {{store_output(("idx_m", "idx_p"), "acc", "mask")}}
+""",
+)
+
+
+B2B_GEMM_PASS = PatternMatcherPass(
+    prevent_match_across_mutations=True,
+    pass_name="b2b_gemm_pass",
+)
+
+
+def can_apply_b2b_gemm(
+    mat1: torch.fx.Node, mat2: torch.fx.Node, mat3: torch.fx.Node
+) -> bool:
+    if not (("val" in mat1.meta) and ("val" in mat2.meta) and ("val" in mat3.meta)):
+        return False
+    mat1 = mat1.meta["val"]
+    mat2 = mat2.meta["val"]
+    mat3 = mat3.meta["val"]
+    if not (mat1.is_cuda and mat2.is_cuda and mat3.is_cuda):
+        return False
+    if not (
+        (len(mat1.shape) == 2) and (len(mat2.shape) == 2) and (len(mat3.shape) == 2)
+    ):
+        return False
+    if not ((mat1.shape[1] == mat2.shape[0]) and (mat2.shape[1] == mat3.shape[0])):
+        return False
+    # TODO: change to a real-check for size restrictions (may consider hardware limit?)
+    m, n, o, p = mat1.shape[0], mat1.shape[1], mat3.shape[0], mat3.shape[1]
+    m_ok = (m % 128 == 0) and (m > 128)
+    n_ok = n == 32
+    o_ok = (o % 128 == 0) and (o > 128)
```

Comment:
Eventually we should handle size that is not multiple of 128 as well

Reply:
Updated to handle arbitrary sizes.